File: MtxVec.Delphi/PersistentStreamer.pas
unit PersistentStreamer;

interface
uses
  Classes;

type
  TPersistentReader = class
  private
    FStream: TStream;
    FMode: TStreamOriginalFormat;
  public
    constructor Create (aStream: TStream; aMode: TStreamOriginalFormat = sofUnknown);
    procedure ReadPersistent(instance: TPersistent);
  end;

  TPersistentWriter = class
  private
    FStream: TStream;
    FMode: TStreamOriginalFormat;
  public
    constructor Create (aStream: TStream; aMode: TStreamOriginalFormat);

    procedure WritePersistent (instance: TPersistent);
  end;

// stream
procedure SavePersistentToStream (anInstance: TPersistent; aStream: TStream; aMode: TStreamOriginalFormat);
procedure LoadPersistentFromStream (anInstance: TPersistent; aStream: TStream; aMode: TStreamOriginalFormat = sofUnknown);

// file
procedure SavePersistentToFile (anInstance: TPersistent; const fileName: string; aMode: TStreamOriginalFormat);
procedure LoadPersistentFromFile (anInstance: TPersistent; const fileName: string; aMode: TStreamOriginalFormat = sofUnknown);

// string
procedure SavePersistentToString (anInstance: TPersistent; out aStr: string; aMode: TStreamOriginalFormat);
procedure LoadPersistentFromString (anInstance: TPersistent; const aStr: string; aMode: TStreamOriginalFormat = sofUnknown);

implementation
uses
  SysUtils;

type
  TPersistentHolder = class (TComponent)
  private
    FInstance: TPersistent;
  published
    property Instance: TPersistent read FInstance write FInstance;
  end;

{ TPersistentReader }

constructor TPersistentReader.Create(aStream: TStream;
  aMode: TStreamOriginalFormat);
begin
  inherited Create;
  FStream:= aStream;
  FMode:= aMode;
end;


procedure TPersistentReader.ReadPersistent(instance: TPersistent);
var
  holder: TPersistentHolder;
  stream: TMemoryStream;
begin
  holder:= TPersistentHolder.Create (nil);
  try
    holder.Instance:= instance;

    if FMode = sofUnknown then
      FMode:= TestStreamFormat (FStream);

    case FMode of
      sofBinary:
        begin
          FStream.ReadComponent (holder);
        end;

      sofText:
        begin
          stream:= TMemoryStream.Create();
          try
            ObjectTextToBinary (FStream, stream);
            stream.Seek(0, soFromBeginning);
            stream.ReadComponent (holder);
          finally
            stream.Free;
          end;
        end;
    end;
  finally
    holder.Free;
  end;
end;

{ TPersistentWriter }

constructor TPersistentWriter.Create(aStream: TStream;
  aMode: TStreamOriginalFormat);
begin
  inherited Create;
  FStream:= aStream;
  FMode:= aMode;
end;


procedure TPersistentWriter.WritePersistent(instance: TPersistent);
var
  holder: TPersistentHolder;
  stream: TMemoryStream;
begin
  holder:= TPersistentHolder.Create (nil);
  try
    holder.Instance:= instance;
    case FMode of
      sofBinary:
        begin
          FStream.WriteComponent (holder);
        end;

      sofText:
        begin
          stream:= TMemoryStream.Create();
          try
            stream.WriteComponent (holder);
            stream.Seek(0, soFromBeginning);
            ObjectBinaryToText (stream, FStream);
          finally
            stream.Free;
          end;
        end;
    end;
  finally
    holder.Free;
  end;
end;

procedure SavePersistentToStream (anInstance: TPersistent; aStream: TStream; aMode: TStreamOriginalFormat);
var
  wr: TPersistentWriter;
begin
  wr:= TPersistentWriter.Create (aStream, aMode);
  try
    wr.WritePersistent (anInstance);
  finally
    wr.Free;
  end;
end;

procedure LoadPersistentFromStream (anInstance: TPersistent; aStream: TStream; aMode: TStreamOriginalFormat);
var
  rd: TPersistentReader;
begin
  rd:= TPersistentReader.Create (aStream, aMode);
  try
    rd.ReadPersistent (anInstance);
  finally
    rd.Free;
  end;
end;

procedure SavePersistentToFile (anInstance: TPersistent; const fileName: string; aMode: TStreamOriginalFormat);
var
  stream: TFileStream;
begin
  stream:= TFileStream.Create (fileName, fmCreate);
  try
    SavePersistentToStream (anInstance, stream, aMode);
  finally
    stream.Free;
  end;
end;

procedure LoadPersistentFromFile (anInstance: TPersistent; const fileName: string; aMode: TStreamOriginalFormat);
var
  stream: TFileStream;
begin
  stream:= TFileStream.Create (fileName, fmOpenRead);
  try
    LoadPersistentFromStream (anInstance, stream, aMode);
  finally
    stream.Free;
  end;
end;

procedure SavePersistentToString (anInstance: TPersistent; out aStr: string; aMode: TStreamOriginalFormat);
var
  stream: TStringStream;
begin
  stream:= TStringStream.Create ('');
  try
    SavePersistentToStream (anInstance, stream, aMode);
    aStr:= stream.DataString;
  finally
    stream.Free;
  end;
end;

procedure LoadPersistentFromString (anInstance: TPersistent; const aStr: string; aMode: TStreamOriginalFormat = sofUnknown);
var
  stream: TStringStream;
begin
  stream:= TStringStream.Create (aStr);
  try
    LoadPersistentFromStream (anInstance, stream, aMode);
  finally
    stream.Free;
  end;
end;

initialization
  RegisterClass (TPersistentHolder);
end.

File: MtxVec.CPP/PersistentStreamer.cpp
//---------------------------------------------------------------------------
#pragma hdrstop

#include "PersistentStreamer.h"

//---------------------------------------------------------------------------
#pragma package(smart_init)
//---------------------------------------------------------------------------

class PersistentHolder: public TComponent {
private:
  TPersistent* instance_;
public:
  __fastcall virtual PersistentHolder(TComponent* aOwner);
__published:
  __property TPersistent* Instance = {read = instance_, write = instance_};
};
//---------------------------------------------------------------------------

__fastcall PersistentHolder::PersistentHolder (TComponent* aOwner)
  : TComponent(aOwner)
{
}
//---------------------------------------------------------------------------

PersistentReader::PersistentReader (TStream& stream, TStreamOriginalFormat mode)
  : fSteram(&stream), fMode(mode)
{
}
//---------------------------------------------------------------------------

void PersistentReader::ReadPersistent(TPersistent& instance)
{
  PersistentHolder* holder = new PersistentHolder (NULL);
  try {
    holder->Instance = &instance;

    if (fMode == sofUnknown)
      fMode = TestStreamFormat (fSteram);

    switch (fMode) {
      case sofBinary:
        fSteram->ReadComponent (holder);
        break;

      case sofText:
        TMemoryStream* stream = new TMemoryStream;
        try {
          ObjectTextToBinary (fSteram, stream);
          stream->Seek(0, soFromBeginning);
          stream->ReadComponent (holder);
        } __finally {
          delete stream;
        }
        break;

    } // switch
  } __finally {
    delete holder;
  }
}
//---------------------------------------------------------------------------

PersistentWriter::PersistentWriter (TStream& stream, TStreamOriginalFormat mode)
  : fSteram(&stream),  fMode(mode)
{
}
//---------------------------------------------------------------------------

void PersistentWriter::WritePersistent (TPersistent& instance)
{
  PersistentHolder* holder = new PersistentHolder (NULL);
  try {
    holder->Instance = &instance;
    switch (fMode) {
      case sofBinary:
        fSteram->WriteComponent (holder);
        break;

      case sofText:
        TMemoryStream* stream = new TMemoryStream();
        try {
          stream->WriteComponent (holder);
          stream->Seek(0, soFromBeginning);
          ObjectBinaryToText (stream, fSteram);
        } __finally {
          delete stream;
        }
        break;

    } // switch

  } __finally {
    delete holder;
  }
}
//---------------------------------------------------------------------------
void SavePersistentToStream (TPersistent& anInstance, TStream& stream, TStreamOriginalFormat mode)
{
  PersistentWriter wr (stream, mode);
  wr.WritePersistent (anInstance);
}
//---------------------------------------------------------------------------

void LoadPersistentFromStream (TPersistent& anInstance, TStream& stream, TStreamOriginalFormat mode)
{
  PersistentReader rd (stream, mode);
  rd.ReadPersistent (anInstance);
}
//---------------------------------------------------------------------------

void SavePersistentToFile (TPersistent& anInstance, const AnsiString& fileName, TStreamOriginalFormat mode)
{
  TFileStream* stream = new TFileStream (fileName, fmCreate);
  try {
    SavePersistentToStream (anInstance, *stream, mode);
  } __finally {
    delete stream;
  }
}
//---------------------------------------------------------------------------

void LoadPersistentFromFile (TPersistent& anInstance, const AnsiString& fileName, TStreamOriginalFormat mode)
{
  TFileStream* stream = new TFileStream (fileName, fmOpenRead);
  try {
    LoadPersistentFromStream (anInstance, *stream, mode);
  } __finally {
    delete stream;
  }
}
//---------------------------------------------------------------------------

void SavePersistentToString (TPersistent& anInstance, AnsiString& aStr, TStreamOriginalFormat mode)
{
  TStringStream* stream = new TStringStream((RawByteString) "");
  try {
    SavePersistentToStream (anInstance, *stream, mode);
    aStr = stream->DataString;
  } __finally {
    delete stream;
  }
}
//---------------------------------------------------------------------------

void LoadPersistentFromString (TPersistent& anInstance, const AnsiString& aStr, TStreamOriginalFormat mode)
{
  TStringStream* stream = new TStringStream ((RawByteString) aStr);
  try {
	LoadPersistentFromStream (anInstance, *stream, mode);
  } __finally {
	delete stream;
  }
}
//---------------------------------------------------------------------------

static void Init()
{
  RegisterClass (__classid(PersistentHolder));
}

#pragma startup Init 64
//---------------------------------------------------------------------------
